language/move-vm/runtime/src/loader.rs

```diff
@@ -182,29 +192,40 @@ impl ModuleCache {
     fn insert(
```

Comment:
Yes, currently we get a new loaded module per pair of link context, compiled module pair, so we will create a new version of the framework for every entry point.  Re: `insert` being recursive under the lock, the main thing that's changed there is before it was recursive outside the lock, and now it's recursive inside the lock -- I'm not sure which is better, but I kind of think the latter, because we are reducing a fair bit of lock traffic this way.  

The cause for concern is of course we will be repeating more work under the lock now.  The VM is not completely frozen in that time -- existing sessions that already have acces to the modules/functions/structs they need can keep going, it's only if you need to load or resolve something new that you will be stuck waiting.  I wouldn't say this situation is ideal, I would characterise it as just workable enough to get things off the ground, but we need to come back to this.  There's two things we can/should do:

- Change the loaded module cache to be keyed by a Merkle hash of its Dependency keys and its own module ID (this avoids re-loading the framework for every package, as a pathological case).  I just want to hold off on doing this straight off the bat, because I haven't convinced myself that hash collisions are not an issue yet.
- Cache function bytecode separately from `LoadedModule` -- currently we copy it into the loaded module but it doesn't change at all, so this is probably the biggest source of waste.  But (a) I don't think this is as important as the first bullet, and (b) we may want to specialise the bytecode in future, so may have to undo that work.

Reply:
my understanding is that every function load (e/g/ `Resolver.function_from_handle` will need the lock (and I think more than just that) so I believe that only code that does not need any resolution will keep running. 
Basically the VM is frozen.
Isn't that your understanding too?

Reply:
Yes, that's my understanding too.